File: 141. Linked List Cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    // hashmap: O(N), 4.33%,
    bool hasCycle(ListNode *head) {
        if(!head) return false;
        unordered_map<ListNode*, int> flag;
        while(head){
            if(flag.find(head) == flag.end())
                flag[head] = 1;
            else return true;
            head = head->next;
        }
        return false;
    }

    // two pointers: O(N), 29.76%
    bool hasCycle(ListNode *head) {
        if(!head) return false;
        ListNode* slow = head;
        ListNode* fast = head;
        // if there's a circle, the faster one will finally "catch" the slower one.
        while(fast && fast->next){    // bug: while(slow && fast)
            slow = slow->next;
            fast = fast->next->next;
            if(slow == fast) return true;
        }
        return false;
    }

};

File: 141. Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    
    # Two pointers: O(N) time, O(1) space
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if head == None: return False
        fast, slow = head, head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if fast == slow: return True
        return False
    
    # hash set: O(N) time & space
    def hasCycle(self, head):
        nodes = set()
        while head != None:
            if head in nodes: 
                return True
            nodes.add(head)
            head = head.next 
        return False

        